File: src/types/product.ts
export interface ProductNutrients {
  calories: number; // kcal
  protein: number; // g
  carbohydrates: number; // g
  sugars: number; // g
  addedSugars?: number; // g
  fat: number; // g
  saturatedFat: number; // g
  fiber: number; // g
  sodium: number; // mg
  salt: number; // g
}

export interface FoodAdditive {
  id: string; // e.g. "e330", "e250"
  name: string; // e.g. "Citric acid", "Sodium nitrite"
  riskLevel?: 'safe' | 'caution' | 'high_risk' | 'unknown';
  function?: string; // e.g. "Preservative", "Emulsifier", "Color"
  description?: string;
}

export interface ProductDetails {
  barcode: string;
  name: string;
  brand: string;
  image?: string;
  thumbnail?: string;
  servingSize?: string;
  servingQuantityGrams?: number;
  nutrients100g: ProductNutrients;
  nutrientsServing?: ProductNutrients;
  ingredientsText?: string;
  ingredientsList?: string[];
  allergens: string[];
  additives: FoodAdditive[];
  novaGroup?: 1 | 2 | 3 | 4; // 1: Unprocessed, 2: Culinary, 3: Processed, 4: Ultra-processed
  nutriScore?: 'a' | 'b' | 'c' | 'd' | 'e';
  ecoScore?: 'a' | 'b' | 'c' | 'd' | 'e';
  categories?: string[];
  labels?: string[];
  source: 'openfoodfacts' | 'cache' | 'sample' | 'manual';
  isComplete: boolean;
  missingFields: string[];
  isEstimated?: boolean;
  estimatedCategory?: string;
  lastUpdated: number;
}

File: android/app/src/main/java/com/nutriscan/app/data/model/Product.kt
package com.nutriscan.app.data.model

import androidx.room.Entity
import androidx.room.PrimaryKey
import com.google.gson.annotations.SerializedName

@Entity(tableName = "products")
data class Product(
    @PrimaryKey
    @SerializedName("barcode")
    val barcode: String,

    @SerializedName("name")
    val name: String,

    @SerializedName("brand")
    val brand: String = "",

    @SerializedName("imageUrl")
    val imageUrl: String? = null,

    @SerializedName("servingSize")
    val servingSize: String = "100 g",

    @SerializedName("servingQuantityGrams")
    val servingQuantityGrams: Double? = null,

    @SerializedName("nutrients100g")
    val nutrients100g: Nutrients = Nutrients(),

    @SerializedName("nutrientsServing")
    val nutrientsServing: Nutrients? = null,

    @SerializedName("ingredientsText")
    val ingredientsText: String? = null,

    @SerializedName("ingredientsList")
    val ingredientsList: List<String> = emptyList(),

    @SerializedName("allergens")
    val allergens: List<String> = emptyList(),

    @SerializedName("additives")
    val additives: List<FoodAdditive> = emptyList(),

    @SerializedName("novaGroup")
    val novaGroup: Int? = null, // 1 to 4

    @SerializedName("nutriScore")
    val nutriScore: String? = null, // "a", "b", "c", "d", "e"

    @SerializedName("categories")
    val categories: List<String> = emptyList(),

    @SerializedName("isComplete")
    val isComplete: Boolean = true,

    @SerializedName("isFavorite")
    val isFavorite: Boolean = false,

    @SerializedName("lastScannedAt")
    val lastScannedAt: Long = System.currentTimeMillis(),

    @SerializedName("qualityScore")
    val qualityScore: Int = 80
)
